From 40cf1c890834c75512962bc2e57c783cb62b92a7 Mon Sep 17 00:00:00 2001
From: Salvo Giangreco <giangrecosalvo9@gmail.com>
Date: Thu, 16 Oct 2025 23:01:37 +0200
Subject: [PATCH] Fallback to Face HIDL 2.0

UN1CA: this patch is not complete! Please check unica/patches/legacy/customize.sh for more info.
---
 .../face/FaceService$FaceServiceWrapper.smali |   4 +-
 .../face/aidl/AidlResponseHandler.smali       |   4 +-
 .../face/aidl/FaceAuthenticationClient.smali  | 195 +++-------
 ...ceProvider$$ExternalSyntheticLambda9.smali |   4 +-
 ...viceExImpl$$ExternalSyntheticLambda5.smali | 135 +------
 .../face/aidl/SemFaceServiceExImpl.smali      |   2 +-
 .../face/aidl/SemFaceSessionCloseClient.smali |   2 +-
 .../hidl/HidlToAidlCallbackConverter.smali    | 189 ++-------
 .../face/hidl/HidlToAidlSensorAdapter.smali   |  14 +-
 .../face/hidl/HidlToAidlSessionAdapter.smali  |  28 +-
 .../sensors/face/hidl/TestHal.smali           | 333 +++++-----------
 .../face/V2_0/ISehBiometricsFace$Proxy.smali  | 360 ++++++++++++++++++
 ...ehBiometricsFaceClientCallback$Proxy.smali | 146 +++++++
 13 files changed, 727 insertions(+), 689 deletions(-)

diff --git a/smali_classes2/vendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy.smali b/smali_classes2/vendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy.smali
--- a/smali_classes2/vendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy.smali
+++ b/smali_classes2/vendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy.smali
@@ -948,6 +948,50 @@
     throw p0
 .end method
 
+.method public final sehCloseTaSession()I
+    .locals 4
+
+    const-string/jumbo v0, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFace"
+
+    invoke-static {v0}, Landroid/hardware/authsecret/V1_0/IAuthSecret$Proxy$$ExternalSyntheticOutline0;->m(Ljava/lang/String;)Landroid/os/HwParcel;
+
+    move-result-object v0
+
+    new-instance v1, Landroid/os/HwParcel;
+
+    invoke-direct {v1}, Landroid/os/HwParcel;-><init>()V
+
+    :try_start_0
+    iget-object p0, p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy;->mRemote:Landroid/os/IHwBinder;
+
+    const/16 v2, 0x12
+
+    const/4 v3, 0x0
+
+    invoke-interface {p0, v2, v0, v1, v3}, Landroid/os/IHwBinder;->transact(ILandroid/os/HwParcel;Landroid/os/HwParcel;I)V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->verifySuccess()V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->releaseTemporaryStorage()V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->readInt32()I
+
+    move-result p0
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    return p0
+
+    :catchall_0
+    move-exception p0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    throw p0
+.end method
+
 .method public final sehGetSecurityLevel(Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$sehGetSecurityLevelCallback;)V
     .locals 4
 
@@ -991,6 +1035,226 @@
     invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
 
     return-void
+
+    :catchall_0
+    move-exception p0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    throw p0
+.end method
+
+.method public final sehGetTaInfo()Ljava/lang/String;
+    .locals 4
+
+    const-string/jumbo v0, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFace"
+
+    invoke-static {v0}, Landroid/hardware/authsecret/V1_0/IAuthSecret$Proxy$$ExternalSyntheticOutline0;->m(Ljava/lang/String;)Landroid/os/HwParcel;
+
+    move-result-object v0
+
+    new-instance v1, Landroid/os/HwParcel;
+
+    invoke-direct {v1}, Landroid/os/HwParcel;-><init>()V
+
+    :try_start_0
+    iget-object p0, p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy;->mRemote:Landroid/os/IHwBinder;
+
+    const/16 v2, 0x16
+
+    const/4 v3, 0x0
+
+    invoke-interface {p0, v2, v0, v1, v3}, Landroid/os/IHwBinder;->transact(ILandroid/os/HwParcel;Landroid/os/HwParcel;I)V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->verifySuccess()V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->releaseTemporaryStorage()V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->readString()Ljava/lang/String;
+
+    move-result-object p0
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    return-object p0
+
+    :catchall_0
+    move-exception p0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    throw p0
+.end method
+
+.method public final sehIsTaSessionClosed()Z
+    .locals 4
+
+    const-string/jumbo v0, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFace"
+
+    invoke-static {v0}, Landroid/hardware/authsecret/V1_0/IAuthSecret$Proxy$$ExternalSyntheticOutline0;->m(Ljava/lang/String;)Landroid/os/HwParcel;
+
+    move-result-object v0
+
+    new-instance v1, Landroid/os/HwParcel;
+
+    invoke-direct {v1}, Landroid/os/HwParcel;-><init>()V
+
+    :try_start_0
+    iget-object p0, p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy;->mRemote:Landroid/os/IHwBinder;
+
+    const/16 v2, 0x13
+
+    const/4 v3, 0x0
+
+    invoke-interface {p0, v2, v0, v1, v3}, Landroid/os/IHwBinder;->transact(ILandroid/os/HwParcel;Landroid/os/HwParcel;I)V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->verifySuccess()V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->releaseTemporaryStorage()V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->readBool()Z
+
+    move-result p0
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    return p0
+
+    :catchall_0
+    move-exception p0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    throw p0
+.end method
+
+.method public final sehOpenTaSession()I
+    .locals 4
+
+    const-string/jumbo v0, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFace"
+
+    invoke-static {v0}, Landroid/hardware/authsecret/V1_0/IAuthSecret$Proxy$$ExternalSyntheticOutline0;->m(Ljava/lang/String;)Landroid/os/HwParcel;
+
+    move-result-object v0
+
+    new-instance v1, Landroid/os/HwParcel;
+
+    invoke-direct {v1}, Landroid/os/HwParcel;-><init>()V
+
+    :try_start_0
+    iget-object p0, p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy;->mRemote:Landroid/os/IHwBinder;
+
+    const/16 v2, 0x11
+
+    const/4 v3, 0x0
+
+    invoke-interface {p0, v2, v0, v1, v3}, Landroid/os/IHwBinder;->transact(ILandroid/os/HwParcel;Landroid/os/HwParcel;I)V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->verifySuccess()V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->releaseTemporaryStorage()V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->readInt32()I
+
+    move-result p0
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    return p0
+
+    :catchall_0
+    move-exception p0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    throw p0
+.end method
+
+.method public final sehPauseEnrollment()I
+    .locals 4
+
+    const-string/jumbo v0, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFace"
+
+    invoke-static {v0}, Landroid/hardware/authsecret/V1_0/IAuthSecret$Proxy$$ExternalSyntheticOutline0;->m(Ljava/lang/String;)Landroid/os/HwParcel;
+
+    move-result-object v0
+
+    new-instance v1, Landroid/os/HwParcel;
+
+    invoke-direct {v1}, Landroid/os/HwParcel;-><init>()V
+
+    :try_start_0
+    iget-object p0, p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy;->mRemote:Landroid/os/IHwBinder;
+
+    const/16 v2, 0x14
+
+    const/4 v3, 0x0
+
+    invoke-interface {p0, v2, v0, v1, v3}, Landroid/os/IHwBinder;->transact(ILandroid/os/HwParcel;Landroid/os/HwParcel;I)V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->verifySuccess()V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->releaseTemporaryStorage()V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->readInt32()I
+
+    move-result p0
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    return p0
+
+    :catchall_0
+    move-exception p0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    throw p0
+.end method
+
+.method public final sehResumeEnrollment()I
+    .locals 4
+
+    const-string/jumbo v0, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFace"
+
+    invoke-static {v0}, Landroid/hardware/authsecret/V1_0/IAuthSecret$Proxy$$ExternalSyntheticOutline0;->m(Ljava/lang/String;)Landroid/os/HwParcel;
+
+    move-result-object v0
+
+    new-instance v1, Landroid/os/HwParcel;
+
+    invoke-direct {v1}, Landroid/os/HwParcel;-><init>()V
+
+    :try_start_0
+    iget-object p0, p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy;->mRemote:Landroid/os/IHwBinder;
+
+    const/16 v2, 0x15
+
+    const/4 v3, 0x0
+
+    invoke-interface {p0, v2, v0, v1, v3}, Landroid/os/IHwBinder;->transact(ILandroid/os/HwParcel;Landroid/os/HwParcel;I)V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->verifySuccess()V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->releaseTemporaryStorage()V
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->readInt32()I
+
+    move-result p0
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    invoke-virtual {v1}, Landroid/os/HwParcel;->release()V
+
+    return p0
 
     :catchall_0
     move-exception p0
@@ -1060,6 +1324,102 @@
     throw p0
 .end method
 
+.method public final sehSetFaceTag(Ljava/util/ArrayList;)I
+    .locals 3
+
+    new-instance v0, Landroid/os/HwParcel;
+
+    invoke-direct {v0}, Landroid/os/HwParcel;-><init>()V
+
+    const-string/jumbo v1, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFace"
+
+    invoke-virtual {v0, v1}, Landroid/os/HwParcel;->writeInterfaceToken(Ljava/lang/String;)V
+
+    const/4 v1, 0x1
+
+    invoke-virtual {v0, v1}, Landroid/os/HwParcel;->writeInt32(I)V
+
+    invoke-virtual {v0, p1}, Landroid/os/HwParcel;->writeInt8Vector(Ljava/util/ArrayList;)V
+
+    new-instance p1, Landroid/os/HwParcel;
+
+    invoke-direct {p1}, Landroid/os/HwParcel;-><init>()V
+
+    :try_start_0
+    iget-object p0, p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy;->mRemote:Landroid/os/IHwBinder;
+
+    const/16 v1, 0x22
+
+    const/4 v2, 0x0
+
+    invoke-interface {p0, v1, v0, p1, v2}, Landroid/os/IHwBinder;->transact(ILandroid/os/HwParcel;Landroid/os/HwParcel;I)V
+
+    invoke-virtual {p1}, Landroid/os/HwParcel;->verifySuccess()V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->releaseTemporaryStorage()V
+
+    invoke-virtual {p1}, Landroid/os/HwParcel;->readInt32()I
+
+    move-result p0
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    invoke-virtual {p1}, Landroid/os/HwParcel;->release()V
+
+    return p0
+
+    :catchall_0
+    move-exception p0
+
+    invoke-virtual {p1}, Landroid/os/HwParcel;->release()V
+
+    throw p0
+.end method
+
+.method public final sehSetRotation(I)I
+    .locals 3
+
+    const-string/jumbo v0, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFace"
+
+    invoke-static {p1, v0}, Landroid/hardware/broadcastradio/V2_0/ITunerSession$Proxy$$ExternalSyntheticOutline0;->m(ILjava/lang/String;)Landroid/os/HwParcel;
+
+    move-result-object p1
+
+    new-instance v0, Landroid/os/HwParcel;
+
+    invoke-direct {v0}, Landroid/os/HwParcel;-><init>()V
+
+    :try_start_0
+    iget-object p0, p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace$Proxy;->mRemote:Landroid/os/IHwBinder;
+
+    const/16 v1, 0x18
+
+    const/4 v2, 0x0
+
+    invoke-interface {p0, v1, p1, v0, v2}, Landroid/os/IHwBinder;->transact(ILandroid/os/HwParcel;Landroid/os/HwParcel;I)V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->verifySuccess()V
+
+    invoke-virtual {p1}, Landroid/os/HwParcel;->releaseTemporaryStorage()V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->readInt32()I
+
+    move-result p0
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->release()V
+
+    return p0
+
+    :catchall_0
+    move-exception p0
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->release()V
+
+    throw p0
+.end method
+
 .method public final setActiveUser(ILjava/lang/String;)I
     .locals 2
 
diff --git a/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter.smali b/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter.smali
--- a/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter.smali
+++ b/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter.smali
@@ -248,11 +248,9 @@
 
     const-string/jumbo v1, "android.hidl.base@1.0::IBase"
 
-    const-string/jumbo v2, "vendor.samsung.hardware.biometrics.face@3.0::ISehBiometricsFaceClientCallback"
-
-    const-string/jumbo v3, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFaceClientCallback"
-
-    filled-new-array {v2, v3, v0, v1}, [Ljava/lang/String;
+    const-string/jumbo v2, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFaceClientCallback"
+
+    filled-new-array {v2, v0, v1}, [Ljava/lang/String;
 
     move-result-object v0
 
@@ -268,7 +266,7 @@
 .method public final interfaceDescriptor()Ljava/lang/String;
     .locals 0
 
-    const-string/jumbo p0, "vendor.samsung.hardware.biometrics.face@3.0::ISehBiometricsFaceClientCallback"
+    const-string/jumbo p0, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFaceClientCallback"
 
     return-object p0
 .end method
@@ -657,7 +655,7 @@
 
     return-void
 
-    :sswitch_7
+    :sswitch_6
     invoke-virtual {p2, v3}, Landroid/os/HwParcel;->enforceInterface(Ljava/lang/String;)V
 
     invoke-virtual {p0}, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->interfaceChain()Ljava/util/ArrayList;
@@ -783,8 +781,8 @@
     invoke-virtual {v3}, Landroid/os/HidlMemory;->dup()Landroid/os/HidlMemory;
 
     move-result-object v3
-    :try_end_1
-    .catch Ljava/io/IOException; {:try_start_1 .. :try_end_1} :catch_1
+    :try_end_0
+    .catch Ljava/io/IOException; {:try_start_0 .. :try_end_0} :catch_0
 
     invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
 
@@ -848,7 +846,7 @@
 
     goto :goto_1
 
-    :cond_7
+    :c_svc_000_4
     invoke-static {v3}, Landroid/os/HidlMemoryUtil;->hidlMemoryToByteArray(Landroid/os/HidlMemory;)[B
 
     move-result-object v3
@@ -882,7 +880,7 @@
 
     return-void
 
-    :catch_1
+    :catch_0
     move-exception v0
 
     new-instance v1, Ljava/lang/RuntimeException;
@@ -891,7 +889,7 @@
 
     throw v1
 
-    :pswitch_2
+    :pswitch_1
     invoke-virtual {p2, v7}, Landroid/os/HwParcel;->enforceInterface(Ljava/lang/String;)V
 
     invoke-virtual {p2}, Landroid/os/HwParcel;->readInt64()J
@@ -1057,7 +1055,7 @@
 
     goto :goto_5
 
-    :cond_b
+    :c_svc_000_8
     invoke-virtual {v1}, Ljava/util/ArrayList;->size()I
 
     move-result v4
@@ -1093,7 +1091,7 @@
 
     goto :goto_3
 
-    :cond_c
+    :c_svc_000_9
     iget-object v1, v0, Lcom/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl;->mHandlerMain:Lcom/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl$1;
 
     new-instance v4, Lcom/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl$$ExternalSyntheticLambda3;
@@ -1116,7 +1114,7 @@
 
     goto :goto_5
 
-    :cond_d
+    :c_svc_000_a
     :goto_4
     const-string/jumbo v0, "sehOnPreviewUpdated: preview data is null or size is 0"
 
@@ -1129,36 +1127,53 @@
 
     return-void
 
+    :pswitch_3
+    invoke-virtual {p2, v10}, Landroid/os/HwParcel;->enforceInterface(Ljava/lang/String;)V
+
+    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt64()J
+
+    move-result-wide v1
+
+    iget-object v0, p0, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->mAidlResponseHandler:Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler;
+
+    invoke-virtual {v0}, Ljava/lang/Object;->getClass()Ljava/lang/Class;
+
+    const-string/jumbo v4, "onLockoutChanged"
+
+    invoke-static {v3, v4}, Landroid/util/Slog;->d(Ljava/lang/String;Ljava/lang/String;)I
+
+    iget-object v3, v0, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler;->mScheduler:Lcom/android/server/biometrics/sensors/BiometricScheduler;
+
+    iget-object v3, v3, Lcom/android/server/biometrics/sensors/BiometricScheduler;->mHandler:Landroid/os/Handler;
+
+    new-instance v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda24;
+
+    invoke-direct {v4}, Ljava/lang/Object;-><init>()V
+
+    iput-object v0, v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda24;->f$0:Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler;
+
+    iput-wide v1, v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda24;->f$1:J
+
+    invoke-static {}, Ljava/lang/invoke/VarHandle;->storeStoreFence()V
+
+    invoke-virtual {v3, v4}, Landroid/os/Handler;->post(Ljava/lang/Runnable;)Z
+
+    return-void
+
     :pswitch_4
     invoke-virtual {p2, v10}, Landroid/os/HwParcel;->enforceInterface(Ljava/lang/String;)V
 
     invoke-virtual {p2}, Landroid/os/HwParcel;->readInt64()J
 
-    move-result-wide v1
-
-    iget-object v0, p0, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->mAidlResponseHandler:Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler;
-
-    invoke-virtual {v0}, Ljava/lang/Object;->getClass()Ljava/lang/Class;
-
-    const-string/jumbo v4, "onLockoutChanged"
-
-    invoke-static {v3, v4}, Landroid/util/Slog;->d(Ljava/lang/String;Ljava/lang/String;)I
-
-    iget-object v3, v0, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler;->mScheduler:Lcom/android/server/biometrics/sensors/BiometricScheduler;
-
-    iget-object v3, v3, Lcom/android/server/biometrics/sensors/BiometricScheduler;->mHandler:Landroid/os/Handler;
-
-    new-instance v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda24;
-
-    invoke-direct {v4}, Ljava/lang/Object;-><init>()V
-
-    iput-object v0, v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda24;->f$0:Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler;
-
-    iput-wide v1, v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda24;->f$1:J
-
-    invoke-static {}, Ljava/lang/invoke/VarHandle;->storeStoreFence()V
-
-    invoke-virtual {v3, v4}, Landroid/os/Handler;->post(Ljava/lang/Runnable;)Z
+    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32Vector()Ljava/util/ArrayList;
+
+    move-result-object v1
+
+    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
+
+    move-result v2
+
+    invoke-virtual {p0, v2, v1}, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->onEnumerate(ILjava/util/ArrayList;)V
 
     return-void
 
@@ -1175,7 +1190,7 @@
 
     move-result v2
 
-    invoke-virtual {p0, v2, v1}, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->onEnumerate(ILjava/util/ArrayList;)V
+    invoke-virtual {p0, v2, v1}, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->onRemoved(ILjava/util/ArrayList;)V
 
     return-void
 
@@ -1184,15 +1199,19 @@
 
     invoke-virtual {p2}, Landroid/os/HwParcel;->readInt64()J
 
-    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32Vector()Ljava/util/ArrayList;
-
-    move-result-object v1
+    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
+
+    move-result v1
 
     invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
 
     move-result v2
 
-    invoke-virtual {p0, v2, v1}, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->onRemoved(ILjava/util/ArrayList;)V
+    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
+
+    move-result v3
+
+    invoke-virtual {p0, v1, v2, v3}, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->onError(III)V
 
     return-void
 
@@ -1203,17 +1222,37 @@
 
     invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
 
+    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
+
     move-result v1
 
     invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
 
     move-result v2
 
-    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
-
-    move-result v3
-
-    invoke-virtual {p0, v1, v2, v3}, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->onError(III)V
+    iget-object v0, p0, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->mAidlResponseHandler:Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler;
+
+    invoke-virtual {v0}, Ljava/lang/Object;->getClass()Ljava/lang/Class;
+
+    const-string/jumbo v4, "onAcquired"
+
+    invoke-static {v3, v4}, Landroid/util/Slog;->d(Ljava/lang/String;Ljava/lang/String;)I
+
+    const-class v3, Lcom/android/server/biometrics/sensors/AcquisitionClient;
+
+    new-instance v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda23;
+
+    invoke-direct {v4}, Ljava/lang/Object;-><init>()V
+
+    iput v1, v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda23;->f$0:I
+
+    iput v2, v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda23;->f$1:I
+
+    invoke-static {}, Ljava/lang/invoke/VarHandle;->storeStoreFence()V
+
+    const/4 v1, 0x0
+
+    invoke-virtual {v0, v3, v4, v1}, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler;->handleResponse(Ljava/lang/Class;Ljava/util/function/Consumer;Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda1;)V
 
     return-void
 
@@ -1222,39 +1261,23 @@
 
     invoke-virtual {p2}, Landroid/os/HwParcel;->readInt64()J
 
-    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
-
-    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
-
-    move-result v1
-
-    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
-
-    move-result v2
-
-    iget-object v0, p0, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->mAidlResponseHandler:Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler;
-
-    invoke-virtual {v0}, Ljava/lang/Object;->getClass()Ljava/lang/Class;
-
-    const-string/jumbo v4, "onAcquired"
-
-    invoke-static {v3, v4}, Landroid/util/Slog;->d(Ljava/lang/String;Ljava/lang/String;)I
-
-    const-class v3, Lcom/android/server/biometrics/sensors/AcquisitionClient;
-
-    new-instance v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda23;
-
-    invoke-direct {v4}, Ljava/lang/Object;-><init>()V
-
-    iput v1, v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda23;->f$0:I
-
-    iput v2, v4, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda23;->f$1:I
-
-    invoke-static {}, Ljava/lang/invoke/VarHandle;->storeStoreFence()V
-
-    const/4 v1, 0x0
-
-    invoke-virtual {v0, v3, v4, v1}, Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler;->handleResponse(Ljava/lang/Class;Ljava/util/function/Consumer;Lcom/android/server/biometrics/sensors/face/aidl/AidlResponseHandler$$ExternalSyntheticLambda1;)V
+    move-result-wide v1
+
+    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
+
+    move-result v3
+
+    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
+
+    move-result v4
+
+    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt8Vector()Ljava/util/ArrayList;
+
+    move-result-object v5
+
+    move-object v0, p0
+
+    invoke-virtual/range {v0 .. v5}, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->onAuthenticated(JIILjava/util/ArrayList;)V
 
     return-void
 
@@ -1263,31 +1286,6 @@
 
     invoke-virtual {p2}, Landroid/os/HwParcel;->readInt64()J
 
-    move-result-wide v1
-
-    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
-
-    move-result v3
-
-    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
-
-    move-result v4
-
-    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt8Vector()Ljava/util/ArrayList;
-
-    move-result-object v5
-
-    move-object v0, p0
-
-    invoke-virtual/range {v0 .. v5}, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;->onAuthenticated(JIILjava/util/ArrayList;)V
-
-    return-void
-
-    :pswitch_a
-    invoke-virtual {p2, v10}, Landroid/os/HwParcel;->enforceInterface(Ljava/lang/String;)V
-
-    invoke-virtual {p2}, Landroid/os/HwParcel;->readInt64()J
-
     invoke-virtual {p2}, Landroid/os/HwParcel;->readInt32()I
 
     move-result v1
@@ -1308,7 +1306,6 @@
 
     :pswitch_data_0
     .packed-switch 0x1
-        :pswitch_a
         :pswitch_9
         :pswitch_8
         :pswitch_7
@@ -1323,9 +1320,8 @@
 
     :sswitch_data_0
     .sparse-switch
-        0xf43484e -> :sswitch_7
-        0xf444247 -> :sswitch_6
-        0xf445343 -> :sswitch_5
+        0xf43484e -> :sswitch_6
+        0xf444247 -> :sswitch_5
         0xf485348 -> :sswitch_4
         0xf494e54 -> :sswitch_3
         0xf504e47 -> :sswitch_2
@@ -1343,7 +1339,7 @@
 .method public final queryLocalInterface(Ljava/lang/String;)Landroid/os/IHwInterface;
     .locals 1
 
-    const-string/jumbo v0, "vendor.samsung.hardware.biometrics.face@3.0::ISehBiometricsFaceClientCallback"
+    const-string/jumbo v0, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFaceClientCallback"
 
     invoke-virtual {v0, p1}, Ljava/lang/String;->equals(Ljava/lang/Object;)Z
 
@@ -1368,7 +1364,7 @@
 .method public final toString()Ljava/lang/String;
     .locals 0
 
-    const-string/jumbo p0, "vendor.samsung.hardware.biometrics.face@3.0::ISehBiometricsFaceClientCallback@Stub"
+    const-string/jumbo p0, "vendor.samsung.hardware.biometrics.face@2.0::ISehBiometricsFaceClientCallback@Stub"
 
     return-object p0
 .end method
diff --git a/smali/com/android/server/biometrics/sensors/face/aidl/SemFaceSessionCloseClient.smali b/smali/com/android/server/biometrics/sensors/face/aidl/SemFaceSessionCloseClient.smali
--- a/smali/com/android/server/biometrics/sensors/face/aidl/SemFaceSessionCloseClient.smali
+++ b/smali/com/android/server/biometrics/sensors/face/aidl/SemFaceSessionCloseClient.smali
@@ -71,7 +71,7 @@
 
     move-result-object v0
 
-    check-cast v0, Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;
+    check-cast v0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;
 
     invoke-interface {v0}, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;->sehCloseTaSession()I
 
diff --git a/smali/com/android/server/biometrics/sensors/face/aidl/FaceProvider$$ExternalSyntheticLambda9.smali b/smali/com/android/server/biometrics/sensors/face/aidl/FaceProvider$$ExternalSyntheticLambda9.smali
--- a/smali/com/android/server/biometrics/sensors/face/aidl/FaceProvider$$ExternalSyntheticLambda9.smali
+++ b/smali/com/android/server/biometrics/sensors/face/aidl/FaceProvider$$ExternalSyntheticLambda9.smali
@@ -491,7 +491,7 @@
 
     move-result-object p0
 
-    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;
+    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;
 
     invoke-interface {p0}, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;->sehOpenTaSession()I
 
@@ -758,7 +758,7 @@
 
     move-result-object p0
 
-    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;
+    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;
 
     invoke-interface {p0}, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;->sehPauseEnrollment()I
 
diff --git a/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlSensorAdapter.smali b/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlSensorAdapter.smali
--- a/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlSensorAdapter.smali
+++ b/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlSensorAdapter.smali
@@ -223,10 +223,6 @@
     invoke-static {v1, v0}, Landroid/util/Slog;->d(Ljava/lang/String;Ljava/lang/String;)I
 
     :try_start_0
-    sget-boolean v0, Lcom/android/server/biometrics/SemBiometricFeature;->FEATURE_JDM_HAL:Z
-
-    if-eqz v0, :cond_2
-
     invoke-static {}, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;->getService()Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;
 
     move-result-object v0
@@ -245,12 +241,6 @@
 
     goto :goto_1
 
-    :cond_2
-    invoke-static {}, Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;->getService()Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;
-
-    move-result-object v0
-
-    iput-object v0, p0, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlSensorAdapter;->mDaemon:Landroid/hardware/biometrics/face/V1_0/IBiometricsFace;
     :try_end_0
     .catch Ljava/util/NoSuchElementException; {:try_start_0 .. :try_end_0} :catch_1
     .catch Landroid/os/RemoteException; {:try_start_0 .. :try_end_0} :catch_0
@@ -272,7 +262,7 @@
     :goto_2
     iget-object v0, p0, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlSensorAdapter;->mDaemon:Landroid/hardware/biometrics/face/V1_0/IBiometricsFace;
 
-    if-nez v0, :cond_3
+    if-nez v0, :c_svc_000_2
 
     const-string p0, "Face HAL not available"
 
@@ -282,7 +272,7 @@
 
     return-object p0
 
-    :cond_3
+    :c_svc_000_2
     invoke-interface {v0}, Landroid/hidl/base/V1_0/IBase;->asBinder()Landroid/os/IHwBinder;
 
     move-result-object v0
diff --git a/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlSessionAdapter.smali b/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlSessionAdapter.smali
--- a/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlSessionAdapter.smali
+++ b/smali/com/android/server/biometrics/sensors/face/hidl/HidlToAidlSessionAdapter.smali
@@ -799,7 +799,7 @@
 
     move-result-object p0
 
-    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;
+    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;
 
     invoke-interface {p0}, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;->sehResumeEnrollment()I
 
@@ -861,7 +861,7 @@
 
     move-result-object p0
 
-    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;
+    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;
 
     invoke-interface {p0, v0}, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;->sehSetFaceTag(Ljava/util/ArrayList;)I
     :try_end_0
@@ -888,10 +888,6 @@
 
     move-result-wide v0
 
-    sget-boolean v2, Lcom/android/server/biometrics/SemBiometricFeature;->FEATURE_JDM_HAL:Z
-
-    if-eqz v2, :cond_0
-
     iget-object v2, p0, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlSessionAdapter;->mSession:Ljava/util/function/Supplier;
 
     invoke-interface {v2}, Ljava/util/function/Supplier;->get()Ljava/lang/Object;
@@ -908,26 +904,6 @@
 
     iget-wide v2, v2, Landroid/hardware/biometrics/face/V1_0/OptionalUint64;->value:J
 
-    goto :goto_0
-
-    :cond_0
-    iget-object v2, p0, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlSessionAdapter;->mSession:Ljava/util/function/Supplier;
-
-    invoke-interface {v2}, Ljava/util/function/Supplier;->get()Ljava/lang/Object;
-
-    move-result-object v2
-
-    check-cast v2, Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;
-
-    iget-object v3, p0, Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlSessionAdapter;->mHidlToAidlCallbackConverter:Lcom/android/server/biometrics/sensors/face/hidl/HidlToAidlCallbackConverter;
-
-    invoke-interface {v2, v3}, Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;->sehSetCallbackEx(Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFaceClientCallback;)Landroid/hardware/biometrics/face/V1_0/OptionalUint64;
-
-    move-result-object v2
-
-    iget-wide v2, v2, Landroid/hardware/biometrics/face/V1_0/OptionalUint64;->value:J
-
-    :goto_0
     new-instance v4, Ljava/lang/StringBuilder;
 
     const-string/jumbo v5, "SetCallback FINISH ("
diff --git a/smali/com/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl.smali b/smali/com/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl.smali
--- a/smali/com/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl.smali
+++ b/smali/com/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl.smali
@@ -1458,7 +1458,7 @@
 
     move-result-object p0
 
-    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;
+    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;
 
     if-eq p1, v11, :cond_3
 
diff --git a/smali/com/android/server/biometrics/sensors/face/aidl/FaceAuthenticationClient.smali b/smali/com/android/server/biometrics/sensors/face/aidl/FaceAuthenticationClient.smali
--- a/smali/com/android/server/biometrics/sensors/face/aidl/FaceAuthenticationClient.smali
+++ b/smali/com/android/server/biometrics/sensors/face/aidl/FaceAuthenticationClient.smali
@@ -895,7 +895,7 @@
 
     sget-boolean v9, Lcom/android/server/biometrics/SemBiometricFeature;->FEATURE_JDM_HAL:Z
 
-    if-nez v9, :cond_3
+    if-nez v9, :c_svc_000_1
 
     iget-object v9, v2, Lcom/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl;->mSensor:Lcom/android/server/biometrics/sensors/face/aidl/Sensor;
 
@@ -1025,10 +1025,10 @@
 
     iput-object v7, v2, Lcom/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl;->mCancellationSignal:Landroid/hardware/biometrics/common/ICancellationSignal;
 
-    goto :goto_2
-
-    :cond_3
-    :goto_1
+    goto :g_svc_000_1
+
+    :c_svc_000_1
+    :g_svc_000_0
     invoke-virtual {v1}, Lcom/android/server/biometrics/sensors/face/aidl/AidlSession;->getSession()Landroid/hardware/biometrics/face/ISession;
 
     move-result-object v1
@@ -1121,11 +1121,11 @@
 
     sput-object v9, Lcom/android/server/biometrics/sensors/face/SemFaceUtils;->mFidoRequestData:[B
 
-    if-nez v13, :cond_7
+    if-nez v13, :c_svc_000_5
 
     new-array v13, v8, [B
 
-    :cond_7
+    :c_svc_000_5
     iget-object v9, v1, Lcom/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl;->mPreviewSurface:Landroid/view/Surface;
 
     check-cast v2, Lvendor/samsung/hardware/biometrics/face/ISehSession$Stub$Proxy;
@@ -1185,14 +1185,14 @@
 
     iput-object v2, v1, Lcom/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl;->mCancellationSignal:Landroid/hardware/biometrics/common/ICancellationSignal;
 
-    goto/16 :goto_5
+    goto/16 :g_svc_000_4
 
     :catchall_0
     move-exception v0
 
-    goto :goto_3
-
-    :cond_8
+    goto :g_svc_000_2
+
+    :c_svc_000_6
     :try_start_1
     new-instance v0, Landroid/os/RemoteException;
 
@@ -1204,14 +1204,14 @@
     :try_end_1
     .catchall {:try_start_1 .. :try_end_1} :catchall_0
 
-    :goto_3
+    :g_svc_000_2
     invoke-virtual {v15}, Landroid/os/Parcel;->recycle()V
 
     invoke-virtual {v14}, Landroid/os/Parcel;->recycle()V
 
     throw v0
 
-    :cond_9
+    :c_svc_000_7
     move/from16 v16, v6
 
     iget-object v2, v1, Lcom/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl;->mISehSession:Lvendor/samsung/hardware/biometrics/face/ISehSession;
@@ -1288,14 +1288,14 @@
 
     iput-object v2, v1, Lcom/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl;->mCancellationSignal:Landroid/hardware/biometrics/common/ICancellationSignal;
 
-    goto :goto_5
+    goto :g_svc_000_4
 
     :catchall_1
     move-exception v0
 
-    goto :goto_4
-
-    :cond_b
+    goto :g_svc_000_3
+
+    :c_svc_000_9
     :try_start_3
     new-instance v0, Landroid/os/RemoteException;
 
@@ -1307,14 +1307,14 @@
     :try_end_3
     .catchall {:try_start_3 .. :try_end_3} :catchall_1
 
-    :goto_4
+    :g_svc_000_3
     invoke-virtual {v14}, Landroid/os/Parcel;->recycle()V
 
     invoke-virtual {v13}, Landroid/os/Parcel;->recycle()V
 
     throw v0
 
-    :cond_c
+    :c_svc_000_a
     move/from16 v16, v6
 
     iget-object v2, v1, Lcom/android/server/biometrics/sensors/face/aidl/SemFaceServiceExImpl;->mContext:Landroid/content/Context;
@@ -1327,11 +1327,11 @@
 
     sput-object v9, Lcom/android/server/biometrics/sensors/face/SemFaceUtils;->mFidoRequestData:[B
 
-    if-nez v6, :cond_d
+    if-nez v6, :c_svc_000_b
 
     new-array v6, v8, [B
 
-    :cond_d
+    :c_svc_000_b
     check-cast v10, Lvendor/samsung/hardware/biometrics/face/ISehSession$Stub$Proxy;
 
     iget-object v9, v10, Lvendor/samsung/hardware/biometrics/face/ISehSession$Stub$Proxy;->mRemote:Landroid/os/IBinder;
@@ -1426,7 +1426,7 @@
 
     iget v6, v1, Lcom/android/server/biometrics/sensors/face/aidl/AidlSession;->mHalInterfaceVersion:I
 
-    if-lt v6, v2, :cond_10
+    if-lt v6, v2, :c_svc_000_e
 
     invoke-virtual {v0}, Lcom/android/server/biometrics/sensors/HalClientMonitor;->getOperationContext()Lcom/android/server/biometrics/log/OperationContextExt;
 
@@ -1458,9 +1458,9 @@
 
     invoke-virtual {v6, v2, v7, v9, v1}, Lcom/android/server/biometrics/log/BiometricContextProvider;->subscribe(Lcom/android/server/biometrics/log/OperationContextExt;Ljava/util/function/Consumer;Ljava/util/function/Consumer;Landroid/hardware/biometrics/AuthenticateOptions;)V
 
-    goto :goto_8
-
-    :cond_10
+    goto :g_svc_000_7
+
+    :c_svc_000_e
     invoke-virtual {v1}, Lcom/android/server/biometrics/sensors/face/aidl/AidlSession;->getSession()Landroid/hardware/biometrics/face/ISession;
 
     move-result-object v1
@@ -1527,7 +1527,7 @@
 
     iput-object v1, v0, Lcom/android/server/biometrics/sensors/face/aidl/FaceAuthenticationClient;->mBackgroundNotification:Lcom/android/server/biometrics/SemBiometricNotification;
 
-    :cond_13
+    :c_svc_000_11
     new-instance v1, Landroid/content/Intent;
 
     const-string/jumbo v6, "com.samsung.android.server.biometrics.BIOMETRICS_NOTIFICATION"
diff --git a/smali_classes2/vendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFaceClientCallback$Proxy.smali b/smali_classes2/vendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFaceClientCallback$Proxy.smali
--- a/smali_classes2/vendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFaceClientCallback$Proxy.smali
+++ b/smali_classes2/vendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFaceClientCallback$Proxy.smali
@@ -364,6 +364,152 @@
     throw p0
 .end method
 
+.method public final onEnumerate(ILjava/util/ArrayList;)V
+    .locals 3
+
+    new-instance v0, Landroid/os/HwParcel;
+
+    invoke-direct {v0}, Landroid/os/HwParcel;-><init>()V
+
+    const-string/jumbo v1, "android.hardware.biometrics.face@1.0::IBiometricsFaceClientCallback"
+
+    invoke-virtual {v0, v1}, Landroid/os/HwParcel;->writeInterfaceToken(Ljava/lang/String;)V
+
+    const-wide/16 v1, 0x0
+
+    invoke-virtual {v0, v1, v2}, Landroid/os/HwParcel;->writeInt64(J)V
+
+    invoke-virtual {v0, p2}, Landroid/os/HwParcel;->writeInt32Vector(Ljava/util/ArrayList;)V
+
+    invoke-virtual {v0, p1}, Landroid/os/HwParcel;->writeInt32(I)V
+
+    new-instance p1, Landroid/os/HwParcel;
+
+    invoke-direct {p1}, Landroid/os/HwParcel;-><init>()V
+
+    :try_start_0
+    iget-object p0, p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFaceClientCallback$Proxy;->mRemote:Landroid/os/IHwBinder;
+
+    const/4 p2, 0x6
+
+    const/4 v1, 0x1
+
+    invoke-interface {p0, p2, v0, p1, v1}, Landroid/os/IHwBinder;->transact(ILandroid/os/HwParcel;Landroid/os/HwParcel;I)V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->releaseTemporaryStorage()V
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    invoke-virtual {p1}, Landroid/os/HwParcel;->release()V
+
+    return-void
+
+    :catchall_0
+    move-exception p0
+
+    invoke-virtual {p1}, Landroid/os/HwParcel;->release()V
+
+    throw p0
+.end method
+
+.method public final onError(III)V
+    .locals 3
+
+    new-instance v0, Landroid/os/HwParcel;
+
+    invoke-direct {v0}, Landroid/os/HwParcel;-><init>()V
+
+    const-string/jumbo v1, "android.hardware.biometrics.face@1.0::IBiometricsFaceClientCallback"
+
+    invoke-virtual {v0, v1}, Landroid/os/HwParcel;->writeInterfaceToken(Ljava/lang/String;)V
+
+    const-wide/16 v1, 0x0
+
+    invoke-virtual {v0, v1, v2}, Landroid/os/HwParcel;->writeInt64(J)V
+
+    invoke-virtual {v0, p1}, Landroid/os/HwParcel;->writeInt32(I)V
+
+    invoke-virtual {v0, p2}, Landroid/os/HwParcel;->writeInt32(I)V
+
+    invoke-virtual {v0, p3}, Landroid/os/HwParcel;->writeInt32(I)V
+
+    new-instance p1, Landroid/os/HwParcel;
+
+    invoke-direct {p1}, Landroid/os/HwParcel;-><init>()V
+
+    :try_start_0
+    iget-object p0, p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFaceClientCallback$Proxy;->mRemote:Landroid/os/IHwBinder;
+
+    const/4 p2, 0x4
+
+    const/4 p3, 0x1
+
+    invoke-interface {p0, p2, v0, p1, p3}, Landroid/os/IHwBinder;->transact(ILandroid/os/HwParcel;Landroid/os/HwParcel;I)V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->releaseTemporaryStorage()V
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    invoke-virtual {p1}, Landroid/os/HwParcel;->release()V
+
+    return-void
+
+    :catchall_0
+    move-exception p0
+
+    invoke-virtual {p1}, Landroid/os/HwParcel;->release()V
+
+    throw p0
+.end method
+
+.method public final onRemoved(ILjava/util/ArrayList;)V
+    .locals 3
+
+    new-instance v0, Landroid/os/HwParcel;
+
+    invoke-direct {v0}, Landroid/os/HwParcel;-><init>()V
+
+    const-string/jumbo v1, "android.hardware.biometrics.face@1.0::IBiometricsFaceClientCallback"
+
+    invoke-virtual {v0, v1}, Landroid/os/HwParcel;->writeInterfaceToken(Ljava/lang/String;)V
+
+    const-wide/16 v1, 0x0
+
+    invoke-virtual {v0, v1, v2}, Landroid/os/HwParcel;->writeInt64(J)V
+
+    invoke-virtual {v0, p2}, Landroid/os/HwParcel;->writeInt32Vector(Ljava/util/ArrayList;)V
+
+    invoke-virtual {v0, p1}, Landroid/os/HwParcel;->writeInt32(I)V
+
+    new-instance p1, Landroid/os/HwParcel;
+
+    invoke-direct {p1}, Landroid/os/HwParcel;-><init>()V
+
+    :try_start_0
+    iget-object p0, p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFaceClientCallback$Proxy;->mRemote:Landroid/os/IHwBinder;
+
+    const/4 p2, 0x5
+
+    const/4 v1, 0x1
+
+    invoke-interface {p0, p2, v0, p1, v1}, Landroid/os/IHwBinder;->transact(ILandroid/os/HwParcel;Landroid/os/HwParcel;I)V
+
+    invoke-virtual {v0}, Landroid/os/HwParcel;->releaseTemporaryStorage()V
+    :try_end_0
+    .catchall {:try_start_0 .. :try_end_0} :catchall_0
+
+    invoke-virtual {p1}, Landroid/os/HwParcel;->release()V
+
+    return-void
+
+    :catchall_0
+    move-exception p0
+
+    invoke-virtual {p1}, Landroid/os/HwParcel;->release()V
+
+    throw p0
+.end method
+
 .method public final ping()V
     .locals 4
 
diff --git a/smali/com/android/server/biometrics/sensors/face/FaceService$FaceServiceWrapper.smali b/smali/com/android/server/biometrics/sensors/face/FaceService$FaceServiceWrapper.smali
--- a/smali/com/android/server/biometrics/sensors/face/FaceService$FaceServiceWrapper.smali
+++ b/smali/com/android/server/biometrics/sensors/face/FaceService$FaceServiceWrapper.smali
@@ -3208,7 +3208,7 @@
 
     move-result-object p0
 
-    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;
+    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;
 
     invoke-interface {p0}, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;->sehGetTaInfo()Ljava/lang/String;
 
@@ -3602,7 +3602,7 @@
 
     move-result-object p0
 
-    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V3_0/ISehBiometricsFace;
+    check-cast p0, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;
+    :g_svc_000_1
+    :g_svc_000_7
+    :g_svc_000_4
 
     invoke-interface {p0}, Lvendor/samsung/hardware/biometrics/face/V2_0/ISehBiometricsFace;->sehIsTaSessionClosed()Z
 
-- 
2.43.0
